Markdownz: Provide backwards compatibility for rendering images with sizes in source string

goplayoutside3 opened the pull request:
_Please request review from `@zooniverse/frontend` team or an individual member of that team._ 

Package:
lib-react-components

Closes #1269.

Describe your changes:
Added a function to `Markdownz` component that manually removes image sizes from their source strings. An image's size is instead only placed in the alt tag so `remark2react` correctly maps markdown to `componentMappings`. This provides backwards compatibility with previously used `markdown-it-imsize` as described in issue #1269. 

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

srallen commented on packages/lib-react-components/src/Markdownz/Markdownz.js:
```diff
@@ -66,6 +66,36 @@ class Markdownz extends React.Component {
 
+  findResizedImages (children) {
+    const imageRegex = /(?:!\[(.*?)\]\((.*?)\))/g
+    const images = Array.from(children.matchAll(imageRegex))
```
Not sure if you saw this or can see this, but github is warning:
> Polynomial regular expression used on uncontrolled data
This regular expression that depends on library input may run slow on strings starting with '!\[' and with many repetitions of '!\[a'.
This regular expression that depends on library input may run slow on strings starting with '!\[\](' and with many repetitions of '\](a'. 

https://github.com/zooniverse/front-end-monorepo/security/code-scanning/29?query=ref%3Arefs%2Fpull%2F2153%2Fhead

So, we need to rethink this particular regex.

srallen replied:
(This may still be the best way to do this, but I'm no regex expert, so will do some research into this)

eatyourgreens replied:
Maybe limit the number of captures this regex can make? I’m not sure of the best solution either but we do know that a markdown string won’t have an infinite number of images.

eatyourgreens replied:
Perhaps add the URL protocol (Question: should we restrict content to https://? I'd vote yes.) and a capture group that grabs the URL. Here's a very quick, rough example: https://regexr.com/5t02b

The URL is captured as group 1, so I think you should be able to get it using backreferences.
![Screenshot of the regexr.com inspector tool, showing a regext which captures only the URL.](https://user-images.githubusercontent.com/59547/118368813-75bf6280-b59a-11eb-9c7a-e1c368083edc.png)

eatyourgreens replied:
Here's an example which splits the capture into groups for the image URL and the size string, which might be more useful. https://regexr.com/5t05e

These examples might still trigger the GitHub warning, though. I'm not 100% sure how to fix that.

![Screenshot of a regex that combines the image and size regexes into a single expression. Examples of the capture group strings are also shown.](https://user-images.githubusercontent.com/59547/118372024-3f3c1480-b5a7-11eb-9d3a-168563eb28af.png)

eatyourgreens replied:
That regex doesn't match this image, so it might still need work.
```markdown
![This is a test image to try out regex matching. It has a couple of sentences in its description.](https://panoptes-uploads.zooniverse.org/production/subject_location/66094a64-8823-4314-8ef4-1ee228e49470.jpeg =100x)
```

eatyourgreens replied:
Changing the second number match to `\d*?` would fix that, I think.

eatyourgreens replied:
That expensive regex is gone from the code now.

goplayoutside3 commented:
@srallen I'm not sure how to solve this regex issue either 😅 I'll do some more research, and maybe I'll need to try a different approach

srallen commented on packages/lib-react-components/src/Markdownz/Markdownz.js:
```diff
@@ -66,6 +66,37 @@ class Markdownz extends React.Component {
+
+    if (images.length) {
+      images.forEach(img => {
+        const imgSizeRegex = /=(\d+(%|px|em|rem|vw)?)x(\d+(%|px|em|rem|vh)?)/
```
This regex will need an adjustment too because `markdown-it-imsize` allows only widths to be set like `=100x`. See how the images are being missed on this branch on TESS: 

<img width="1003" alt="Screen Shot 2021-05-14 at 3 10 03 PM" src="https://user-images.githubusercontent.com/5016731/118323646-89999480-b4c6-11eb-908d-50a7a7695063.png">

Here's the markdown string to work with:
`![exoplanet_cropped.jpg](https://panoptes-uploads.zooniverse.org/production/project_attached_image/38c81b00-5970-468b-9a91-8e0d102a8e82.jpeg =300x)`

I think adding `?` at the end should fix it: `/=(\d+(%|px|em|rem|vw)?)x(\d+(%|px|em|rem|vh)?)?/` but we should add this to the example markdown and tests to be sure.

eatyourgreens replied:
Could this be placed inside `renderMedia()`, to transform `nodeProps.src`?

`\d+` might be too greedy. A real image size won’t have an infinite number of digits so you could limit the search to maybe 3 or 4 digits, max.

eatyourgreens replied:
Just caught up on the conversation elsewhere. So this can’t go in `renderMedia` because the URL is validated before that’s invoked?

goplayoutside3 replied:
Correct. `remark` catches the image url with resize before as an invalid string before `renderMedia` can be invoked, as referenced here #1269.

srallen reviewed (changes requested):
I did one more thorough review of this and I also have a suggestion fo the image size regex itself. Still unsure what to do about Github's warning, though.

eatyourgreens commented on packages/lib-react-components/.storybook/lib/example.md:
```diff
@@ -6,7 +6,8 @@
 |relative link|`[relative-link](/relative)`|[relative-link](/relative)|
 |image|`![imagealttext](https://panoptes-uploads.zooniverse.org/production/subject_location/66094a64-8823-4314-8ef4-1ee228e49470.jpeg)`|![imagealttext](https://panoptes-uploads.zooniverse.org/production/subject_location/66094a64-8823-4314-8ef4-1ee228e49470.jpeg)|
-|image resized|`![imagealttext =100x100](https://panoptes-uploads.zooniverse.org/production/subject_location/66094a64-8823-4314-8ef4-1ee228e49470.jpeg)`|![imagealttext =100x100](https://panoptes-uploads.zooniverse.org/production/subject_location/66094a64-8823-4314-8ef4-1ee228e49470.jpeg)|
+|image resized in alt text|`![imagealttext =100x100](https://panoptes-uploads.zooniverse.org/production/subject_location/66094a64-8823-4314-8ef4-1ee228e49470.jpeg)`|![imagealttext =100x100](https://panoptes-uploads.zooniverse.org/production/subject_location/66094a64-8823-4314-8ef4-1ee228e49470.jpeg)|
```
I assume this will be OK, but what happens if my alt text contains an `=` character? eg. 
```html
<img alt='A blackboard showing the expression 2x2=4' … />
```

Maybe use realistic alt text strings in the test cases, even if it's just lorem ipsum?

goplayoutside3 replied:
I'll add this example to the storybook example.md.

eatyourgreens commented:
You're scanning the whole `newChildren` string once for each matched image string, which probably isn't efficient for large strings. 

It might be more efficient to scan `children` once and replace each image as it's matched. Here, `replaceImage` would be called once for each match and returns the substitution for that match. (I think I've got the syntax right for capture groups but I'm not 100% sure.)

```js
function replaceImage(...args) {
  const [ match, altText, imageURL, imageSize, ...rest ] = args
  return `![${altText} ${imageSize}](${imageURL})`
}

const imageRegex = /!\[([^\]]+?)\]\((https:\/\/[^)]+?) (=\d+?(|%|px|em|rem|vw)x\d+?(|%|px|em|rem|vh))\)/g

const newChildren = children.replace(imageRegex, replaceImage)
```

eatyourgreens commented:
Something else that occurred to me. This expression won't match images that have no text description `![](https://example.domain/photo.jpg =100x)`. I think that's ok, because we want to encourage people to make their images accessible. I've added that to the list of test examples on the regexr page.

eatyourgreens commented on packages/lib-react-components/src/Markdownz/Markdownz.js:
```diff
@@ -75,6 +106,8 @@ class Markdownz extends React.Component {
       console.warn('Overriding the rendering function for the img tag may break the syntax support for image resizing and using image markup for video and audio. Are you sure you want to do this?')
     }
 
+    const childrenWIthImageResize = this.findResizedImages(children)
```
typo
```suggestion
    const childrenWithImageResize = this.findResizedImages(children)
```

eatyourgreens replied:
Does the regex need to be run on every render? It seems like it's only needed when the component first renders and, optionally, if the `children` string changes.

goplayoutside3 replied:
Agreed! I'd like to save this for a future PR though.

eatyourgreens replied:
Yes! If `remark()` is anything like `markdown-it` then setting it up from scratch on every render might be expensive too. Definitely outside the scope of this PR.

eatyourgreens commented on packages/lib-react-components/src/Markdownz/Markdownz.js:
```diff
@@ -111,7 +144,7 @@ class Markdownz extends React.Component {
       .use(toc)
       .use(remark2react, { remarkReactComponents })
-      .processSync(children).contents
+      .processSync(childrenWIthImageResize).contents
```
typo
```suggestion
      .processSync(childrenWithImageResize).contents
```

goplayoutside3 commented:
> I did one more thorough review of this and I also have a suggestion fo the image size regex itself. Still unsure what to do about Github's warning, though.

I implemented this suggested change to `imgSizeRegex` in `findResizedImages` and `renderMedia`, and I added a spec for it. The TESS formatted image urls now render correctly in storybook.

goplayoutside3 commented:
> You're scanning the whole `newChildren` string once for each matched image string, which probably isn't efficient for large strings.
> 
> It might be more efficient to scan `children` once and replace each image as it's matched. Here, `replaceImage` would be called once for each match and returns the substitution for that match. (I think I've got the syntax right for capture groups but I'm not 100% sure.)

@eatyourgreens I refactored my latest commit to to use `children.replace()` as you suggested. Let me know if that's what you were thinking!

eatyourgreens commented:
@goplayoutside3 that looks great. I have one more suggestion, which is to remove a redundant regex inside the `replaceImageString` method. That method will receive the matched string, plus all matched capture groups, as arguments. You can take advantage of that to get the alt text, URL and size string directly from the method arguments. I'll try that locally, to remind myself how the syntax works, then post a code suggestion.

eatyourgreens commented on packages/lib-react-components/src/Markdownz/Markdownz.js:
```diff
@@ -60,12 +60,33 @@ class Markdownz extends React.Component {
     return null
   }
 
+  replaceImageString (img) {
+    const imgSizeRegex = /=(\d+(%|px|em|rem|vw)?)x(\d+(%|px|em|rem|vh)?)?/
+    const imgSizeMatch = img.match(imgSizeRegex)
+
+    let newImgStr = ''
+
+    if (imgSizeMatch && imgSizeMatch.length > 0) {
+      // delete imgSizeMatch from alt and url string (including the preceding space character)
+      const sanitizedImgStr = img.replace(` ${imgSizeMatch[0]}`, '')
+
+      // place imgSizeMatch into only alt tag
+      const altTagRegex = /(?:!\[(.*?)\])/
+      const altTag = sanitizedImgStr.match(altTagRegex)
+      const altTextWithSize = `${altTag[1]} ${imgSizeMatch[0]}`
+
+      newImgStr = sanitizedImgStr.replace(altTag[0], `![${altTextWithSize}]`)
+    }
+    return newImgStr
+  }
```
`imageRegex`, in the render method, includes regexes that capture the alt text, image URL and size. Those get passed as additional arguments to the substitution function, so you can rewrite this whole method as the following. That should remove the overhead of running some expensive regexes twice.
```suggestion
  replaceImageString (img, altText, imageURL, imageSize) {
    return `![${altText} ${imageSize}](${imageURL})`
  }
```

eatyourgreens commented on packages/lib-react-components/src/Markdownz/Markdownz.js:
```diff
@@ -51,7 +51,7 @@ class Markdownz extends React.Component {
   renderMedia (nodeProps) {
     let width, height
-    const imgSizeRegex = /=(\d+(%|px|em|rem|vw)?)x(\d+(%|px|em|rem|vh)?)/
+    const imgSizeRegex = /=(\d+(%|px|em|rem|vw)?)x(\d+(%|px|em|rem|vh)?)?/
```
Use lazy matching here when matching digits, and explicitly specify nothing as an option for the units.
```suggestion
    const imgSizeRegex = /=(\d+?(|%|px|em|rem|vw))x(\d*?(|%|px|em|rem|vh))/
```

goplayoutside3 replied:
This suggested `imgSizeRegex` does not work correctly in the case of `=100x100` in the alt tag of an image, the match captures and returns only `=100x` instead of both width and height.

eatyourgreens replied:
D’oh! I thought I’d copied this straight from the regex in the render method.

eatyourgreens replied:
I forgot to anchor the expression to the end of the alt text string. This should work, but isn't a required change at all.
```suggestion
    const imgSizeRegex = /=(\d+?)(|%|px|em|rem|vw)x(\d*?)(|%|px|em|rem|vh)$/
```

Try it out here, with tests:  https://regexr.com/5t7uq

eatyourgreens reviewed (changes requested):
I've got a couple of suggested changes to speed up the regex matching, and hopefully fix GitHub security warnings.

eatyourgreens commented:
If those changes don't fix the GitHub warning, I've got one more suggestion that adds limits to the number of digits matched. Let's see what happens, though, because the image regex is already quite complicated.

```js
const imageRegex = /!\[([^\]]+?)\]\((https:\/\/.+?) (=\d{1,3}(|%|px|em|rem|vw)x\d{0,3}(|%|px|em|rem|vh))\)/g
```

goplayoutside3 commented:
@eatyourgreens @srallen  I think the github regex warning as been solved by the last couple of commits. 

I have a follow-up question about styling in `ThumbnailImage`. The `renderMedia` function we're working with in `Markdownz` supports a variety of height x width units, but the max-height and max-width of `ThumbnailImage` are hardcoded to px.

https://github.com/zooniverse/front-end-monorepo/blob/03b0c9a6dac4607aadf0e74facc689bac9fbbdf1/packages/lib-react-components/src/Media/components/ThumbnailImage/ThumbnailImage.js#L11

I see that maxWidth and maxHeight are also defined directly in the `StyledBox` component in `ThumbnailImage`. Is that intentional? 

I'd like to handle cases such as TESS images where there's only a width defined such as `![alttext =100x](etc)` and set maxHeight to none or maybe inherit.

eatyourgreens commented:
I’m not seeing the security warning any more. Changes look good to me. I’ll give it another review in the morning.

eatyourgreens reviewed (approved):
This looks great! The performance issues seems to have been solved by removing the greedy image matching from `replaceImageString`. I've got a suggestion for matching the alt text in `renderMedia` but feel free to ignore it.

eatyourgreens commented:
If anyone has specified fixed height images in their markdown, this PR won't catch them.
```markdown
![This is a fixed height image which adjusts width to fit.](https://panoptes-uploads.zooniverse.org/production/subject_location/66094a64-8823-4314-8ef4-1ee228e49470.jpeg =x100)
```

I wouldn't worry about that unless it comes up in an actual project.

srallen reviewed (approved):
Great work! PH TESS about page is rendering images:

<img width="379" alt="Screen Shot 2021-05-19 at 1 32 11 PM" src="https://user-images.githubusercontent.com/5016731/118865887-0ea9f280-b8a7-11eb-8d80-73f87b62d370.png">

And the images in Galaxy Zoo's task:

<img width="391" alt="Screen Shot 2021-05-19 at 1 33 22 PM" src="https://user-images.githubusercontent.com/5016731/118865952-1e293b80-b8a7-11eb-9ff5-47a202458ae2.png">

So now we can fix how these are styled.

About the `ThumbnailImage` hardcoding to px, could you open an issue for it? I don't think that needs to be solved in this PR.

goplayoutside3 commented:
Sounds good! I opened #2189 to address ThumbnailImage dimensions.
